client/notices.go

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) 2023 Canonical Ltd
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 as
+// published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package client
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/url"
+	"time"
+)
+
+type NotifyOptions struct {
+	// Key is the custom notice's key. Must be in "domain.com/key" format.
+	Key string
+
+	// Data are optional key=value pairs for this occurrence of the notice.
+	Data map[string]string
+
+	// RepeatAfter allows the notice to repeat after this duration. If this is
+	// zero (the default), the notice will always repeat.
+	RepeatAfter time.Duration
+}
+
+// Notify records an occurrence of a "custom" notice with the specified options,
+// returning the notice ID.
+func (client *Client) Notify(opts *NotifyOptions) (string, error) {
+	var payload = struct {
+		Action      string            `json:"action"`
+		Type        string            `json:"type"`
+		Key         string            `json:"key"`
+		RepeatAfter string            `json:"repeat-after,omitempty"`
+		Data        map[string]string `json:"data,omitempty"`
+	}{
+		Action: "add",
+		Type:   "custom",
+		Key:    opts.Key,
+		Data:   opts.Data,
+	}
+	if opts.RepeatAfter != 0 {
+		payload.RepeatAfter = opts.RepeatAfter.String()
+	}
+	var body bytes.Buffer
+	if err := json.NewEncoder(&body).Encode(&payload); err != nil {
+		return "", err
+	}
+
+	result := struct {
+		ID string `json:"id"`
+	}{}
+	_, err := client.doSync("POST", "/v1/notices", nil, nil, &body, &result)
+	if err != nil {
+		return "", err
+	}
+	return result.ID, err
+}
+
+type NoticesOptions struct {
+	// Types, if not empty, includes only notices whose type is one of these.
+	Types []NoticeType
+
+	// Keys, if not empty, includes only notices whose key is one of these.
+	Keys []string
+
+	// After, if set, includes only notices that were last repeated after this time.
+	After time.Time
+}
+
+// Notice is a notification whose identity is the combination of Type and Key
+// that has occurred Occurrences number of times.
```

Comment:
Perhaps something along the lines of:
```
// Notice holds details of an event that was observed and reported either inside
// the server itself or externally via the API. Besides the ID field itself, the Type
// and Key fields together also uniquely identify a particular notice, and when a new
// notification is made with matching Type and Key, the previous notice is updated
// appropriately instead of a new one being created.
```
This intro should hopefully make some of these fields more obvious.

Reply:
Done.